### PR TITLE
Implementing Eig Function To paddle.Tensor - To Fix

### LINKED ISSUE #17958: eig
Implementing the [eig](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/linalg/eig_en.html) frontend function from the Paddlepaddle framework.

### PR DESCRIPTION
**Pull request by peps212**
Close #17958 

Hi! I'm trying to implement the [eig frontend function ](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/linalg/eig_en.html), but I'm running into some errors when running the tests. 

For the sake of Iterating quickly, I've limited the shape of the input matrix only to [2,2], I'll add additional cases when I'll figure out how to solve the issue.

The error I get is an assertion error: 

> AssertionError:  the results from backend jax and ground truth framework paddle do not match
> E            [[0.20412415-0.54006172j 0.20412415+0.54006172j]
> E            [0.81649658+0.j         0.81649658-0.j        ]]!=[[-0.20412415+0.54006172j -0.20412415-0.54006172j]
> E            [-0.81649658+0.j         -0.81649658-0.j        ]] 

Plus, the datatype of the two matrices is different, one has a _complex64_ datatype, the other _complex128_.

I started digging into other Ivy's files and found another eig testing function being implemented in [ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_paddle_linalg.py](https://github.com/unifyai/ivy/blob/master/ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_paddle_linalg.py) [Line 282]. What I've noticed here is that this test runs into the same problem I'm facing, the only difference is that this test function has `Test_values=False` so it runs without calling out any errors. 

If i set `Test_values=False` in my test function then I get no errors too, but I think that might defy the purpose of testing. This is my first time contributing to Ivy, so there might be something I'm not seeing, could someone help me figure this out? 

Thanks!

### CODE REVIEW

**Comment by nathzi1505** [comment]
Hi @hmahmood24, would you mind looking at this?

**Comment by hmahmood24** [comment]
> Hi @hmahmood24, would you mind looking at this?

Sure @nathzi1505, I'll look into it over the weekend. Thanks both :)

File: ivy/functional/frontends/paddle/tensor/tensor.py [hunk 1]
```diff
@@ -240,3 +244,10 @@ def astype(self, dtype):
     )
     def logical_xor(self, y, out=None, name=None):
         return paddle_frontend.logical_xor(self, y, out=out)
+
+    @with_supported_device_and_dtypes(
+        {"2.5.0 and below": {"cpu": ("float32", "float64", "complex128", "complex64")}},
+        "paddle",
+    )
```
Comment (hmahmood24) [inline ivy/functional/frontends/paddle/tensor/tensor.py:251]: I think just the `@with_supported_dtypes` decorator would suffice here

File: ivy/functional/frontends/paddle/tensor/tensor.py [hunk 2]
```diff
@@ -240,3 +244,10 @@ def astype(self, dtype):
     )
     def logical_xor(self, y, out=None, name=None):
         return paddle_frontend.logical_xor(self, y, out=out)
+
+    @with_supported_device_and_dtypes(
+        {"2.5.0 and below": {"cpu": ("float32", "float64", "complex128", "complex64")}},
+        "paddle",
+    )
+    def eig(self):
+        return ivy.eig(self._ivy_array)
```
Comment (hmahmood24) [inline ivy/functional/frontends/paddle/tensor/tensor.py:253]: There is already an implementation of `eig` in the `paddle` frontend. Kindly use that here

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_paddle_tensor.py [hunk 3]
```diff
@@ -1316,3 +1316,34 @@ def test_paddle_logical_xor(
...
+# eig
+@handle_frontend_method(
+    class_tree=CLASS_TREE,
+    init_tree="paddle.to_tensor",
+    method_name="eig",
+    dtypes_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("valid"), shape=[2, 2]
+    ),
+)
+def test_paddle_eig(
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_paddle_tensor.py:1330]: `test_paddle_instance_eig` here

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
Hey @peps212. The issue you are facing is an inherent issue with the way `tensorflow` has implemented its `eig` function. It differs from the rest of the backend frameworks and returns  a very different result as flagged [here](https://github.com/tensorflow/tensorflow/issues/60590) on their own repository. Regarding the `test_values` flag, I think @zaeemansari70 would be in a better position to suggest the best course of action considering this is the `linalg` submodule. Other than that, requested a minor changes in the existing PR. Thanks :)

**Comment by peps212** [comment]
Hi @hmahmood24, thank you for the code review. I now understand how the issue is caused by how tensorflow has implemented their eig function, thanks for pointing that out. I made the changes you requested and set `test_values=False`, now the test runs without any error. 

Also, I added an import on top of the Test_paddle_function.py, I'm using the `_get_dtype_and_square_matrix ` function that has been implemented in the test_paddle_linalg.py, which is exactly the function that generates the data for the other eig function. I decided to use that to stay consistent with the other code, I hope is not a problem. If it is i can revert the change.

File: ivy/functional/frontends/paddle/tensor/tensor.py (hunk 2 above)
Comment (peps212) [reply ivy/functional/frontends/paddle/tensor/tensor.py:252]: Maybe i'm getting this wrong, but when calling the ivy.eig() i'm using the eig function defined in the linear_algebra file, I think you might be referring to that. Is this not what you mean?

Comment (zaeemansari70) [reply ivy/functional/frontends/paddle/tensor/tensor.py:252]: What @hmahmood24 is referring to is use something like `paddle_frontend.eig(self)` as this is an instance method

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_paddle_tensor.py [hunk 4]
```diff
@@ -1316,3 +1319,33 @@ def test_paddle_logical_xor(
...
+        init_input_dtypes=input_dtype,
+        init_all_as_kwargs_np={"data": x},
+        method_input_dtypes=input_dtype,
+        method_all_as_kwargs_np={},
+        frontend_method_data=frontend_method_data,
+        init_flags=init_flags,
+        method_flags=method_flags,
+        frontend=frontend,
+        on_device=on_device,
+        test_values=False,
```
Comment (zaeemansari70) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_paddle_tensor.py:1350]: You have to test it by reproducing the input from the output as the issue with some linalg functions is that ther have multiple correct values as a solution.
A sample would be as follows:
https://github.com/unifyai/ivy/blob/ca5c9aa8bb5d30fd4591fc338fe27329bcf61fcc/ivy_tests/test_ivy/test_frontends/test_jax/test_lax/test_linalg.py#L177

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Sorry for the delayed review.
Let me know if you have any questions, Thanks! :slightly_smiling_face:

**Comment by nathzi1505** [comment]
Hi @peps212, looks like there are merge conflicts in the commits you made.
Would you mind revising them?

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Pull request closed without merging**